coreaudio.c

```diff
@@ -0,0 +1,857 @@
+static AudioDeviceID coreaudio_device_id;
+static AudioDeviceIOProcID coreaudio_proc_id = NULL;
+static AudioStreamBasicDescription coreaudio_format_description ;
+static AudioUnit coreaudio_audio_unit = NULL;
+static char *coreaudio_device_name = NULL;
+static bool coreaudio_enable_hog_mode = false;
+static bool coreaudio_use_direct_output = false;
+static UInt32 coreaudio_buffer_frame_size = 1;
+static coreaudio_ring_buffer_t coreaudio_ring_buffer;
+
+
+static OSStatus play_callback(
+	void *inRefCon, AudioUnitRenderActionFlags *ioActionFlags,
+        const AudioTimeStamp *inTimeStamp, UInt32 inBusNumber,
+        UInt32 inNumberFrames, AudioBufferList *ioData)
+{
+	int count = inNumberFrames * coreaudio_format_description.mBytesPerFrame;
+	ioData->mBuffers[0].mDataByteSize = coreaudio_ring_buffer_read(&coreaudio_ring_buffer, ioData->mBuffers[0].mData, count);
+ 	return noErr;
+}
+
+static OSStatus direct_callback(AudioDeviceID inDevice,
+	const AudioTimeStamp *inNow,
+	const AudioBufferList *inInputData,
+	const AudioTimeStamp *inInputTime,
+	AudioBufferList *outOutputData,
+	const AudioTimeStamp *inOutputTime,
+	void *inClientData)
+{
+	int count = outOutputData->mBuffers[0].mDataByteSize;
+	outOutputData->mBuffers[0].mDataByteSize = coreaudio_ring_buffer_read(&coreaudio_ring_buffer, outOutputData->mBuffers[0].mData, count);
+	return noErr;
+}
+
+static int coreaudio_init(void)
+{
+	coreaudio_device_id = kAudioDeviceUnknown;
+	return OP_ERROR_SUCCESS;
+}
+
+static int coreaudio_exit(void)
+{
+	AudioHardwareUnload();
+	return OP_ERROR_SUCCESS;
+}
+
+static int coreaudio_close(void)
+{
+	if (coreaudio_use_direct_output)
+	{
+		AudioOutputUnitStop(coreaudio_audio_unit);
+		AudioUnitUninitialize(coreaudio_audio_unit);
+		AudioComponentInstanceDispose(coreaudio_audio_unit);
+	}
+	else
+	{
+		AudioDeviceDestroyIOProcID(coreaudio_device_id, coreaudio_proc_id);
+	}
+	coreaudio_proc_id = NULL;
+	coreaudio_device_id = kAudioDeviceUnknown;
+	coreaudio_ring_buffer_destroy(&coreaudio_ring_buffer);
+	return OP_ERROR_SUCCESS;
+}
+
+static int coreaudio_open(sample_format_t sf, const channel_position_t *channel_map)
+{
+	OSStatus err;
+	UInt32 property_size;
+	Boolean found_device = false;
+
+	AudioObjectPropertyAddress property_address = {
+		kAudioHardwarePropertyDefaultOutputDevice,
+		kAudioObjectPropertyScopeOutput,
+		kAudioObjectPropertyElementMaster
+	};
+
+  	// Get the default output device index as a fallback.
+	UInt32 coreaudio_device_id_size = sizeof(coreaudio_device_id);
+	err = AudioObjectGetPropertyData(
+		kAudioObjectSystemObject, &property_address, 0, NULL,
+		&coreaudio_device_id_size, &coreaudio_device_id);
+	if (err != noErr) {
+		errno = ENODEV;
+		return -OP_ERROR_ERRNO;
+	}
+
+	if (coreaudio_device_name)
+	{
+		property_address.mSelector = kAudioHardwarePropertyDevices;
+		err = AudioObjectGetPropertyDataSize(
+			kAudioObjectSystemObject,
+			&property_address, 0, NULL, &property_size);
+		if (err == noErr)
```

Comment:
Doesn't matter. We then fallback to use the default system output. See line 309. We can do a d_print, though.